### PR TITLE
Subclient document

### PR DESCRIPTION
**Pull request by chunyu3**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Comment by annelo-msft** [comment]
Would you please link the PR to the issue by putting "Fixes " in the issue description?

File: doc/Data Plane Code Generation/SubClients.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# Customizing the generated code
+
+## SubClients
+
+There are two categories of clients: service clients and their subclients. Service clients can be instantiated and have the Client suffix. Subclients can only be created by calling factory methods on other clients (commonly on service clients) and do not have the client suffix. The service client is the entry point to the API for an Azure service.
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/SubClients.md:5]: Please reference the guidelines in this section.

File: doc/Data Plane Code Generation/SubClients.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+There are two approaches to organize subclients and service client.
+
+- **Single top-level client**: define one top-level service client with the name $"{language.default.name}Client", all other clients will be its sub-client.
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/SubClients.md:9]: What does $"{language.default.name}" signify?  This is not a naming convention I'm familiar with, and I'm not sure service team partners will be either.  Would it be possible to use a different convention, maybe  [`<service_name>Client`](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-client) like is used in the guidelines?

Comment (annelo-msft) [inline doc/Data Plane Code Generation/SubClients.md:9]: nit:
```suggestion
- **Single top-level client**: define one top-level service client with the name $"{language.default.name}Client", all other clients will be its subclient.
```

File: doc/Data Plane Code Generation/SubClients.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
+# Customizing the generated code
+
+## SubClients
+
+There are two categories of clients: service clients and their subclients. Service clients can be instantiated and have the Client suffix. Subclients can only be created by calling factory methods on other clients (commonly on service clients) and do not have the client suffix. The service client is the entry point to the API for an Azure service.
+
+There are two approaches to organize subclients and service client.
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/SubClients.md:7]: Single top-level client and client hierarchy are not mutually exclusive -- it would be reasonable for a subclient organization that reflects a client hierarchy to have a single top-level client (like `BlobServiceClient` in the blob storage client hierary.

**Review by annelo-msft — COMMENTED** [review]
Thanks for getting this started, @chunyu3!

I have a few high-level thoughts initially:
- One of the reason the generator customizations samples work is that they are factual and not opinionated; i.e. they tell you how to do things, but not whether or not you should do them.  It would be useful to have this document more like that:  With the DPG tool, you can do A) and B).  To do A, do this.  To do B, do this.  See the guidelines for why you would do A and/or B.
- In the generator customizations samples, the examples are very short, for example: 
image.  For service teams using this guidance, it might be easier to see what's happening if we showed the APIView (like this example does in the "Generated Code After" view, so they can see the client relationships clearly.  It can be hard to see those in a long block of code.  It would also be great if the names of the clients could indicate their relationships to each other.

Happy to help you brainstorm on this if you'd like!

File: doc/Data Plane Code Generation/SubClients.md (hunk 3 above)
Comment (annelo-msft) [inline doc/Data Plane Code Generation/SubClients.md:7]: Maybe if we said instead of "two approaches" that we give you two tools to configure how your service clients and subclients are related: the first lets you specify which is the top-level client, and the other lets you specify parent-child relationships between the subclients.  The latter allows you to make the client hierarchy deeper.  You can use them together.  Suggestions for when to add a subclient are here: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-choosing-client-types

Comment (chunyu3) [reply doc/Data Plane Code Generation/SubClients.md:7]: yes, customer can combine those two. But shall we add those into document, it make it complex and confuse customer. How about add a Note behind this two approach to indicate this combine and set BlobServiceClient as an example?

Comment (chunyu3) [reply doc/Data Plane Code Generation/SubClients.md]: updated, please help to review.

**Comment by chunyu3** [comment]
@annelo-msft I have updated the doc, would you please take a look at it? Thanks

File: doc/Data Plane Code Generation/Subclients.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+### AutoRest Configuration
+> see https://aka.ms/autorest
+
+``` yaml
+namespace: Azure.Service.SubClients
+input-file: $(this-folder)/SubClients.json
+data-plane: true
+security: AzureKey
+security-header-name: Fake-Subscription-Key
+single-top-level-client: true
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/Subclients.md:55]: It would be good to highlight that this is the setting that adds the client and not the others.  Is there a way to specify the name of the top-level client?

File: doc/Data Plane Code Generation/Subclients.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            return Volatile.Read(ref _cachedParameter) ?? Interlocked.CompareExchange(ref _cachedParameter, new Parameter(ClientDiagnostics, _pipeline, _keyCredential, _endpoint), null) ?? _cachedParameter;
+        }
+    }
+}
+
+// SubClient: Root Generated\Root.cs
+namespace Azure.Service.SubClients
+{
+-   public partial class RootClient
++   public partial class Root
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/Subclients.md:125]: Typically `root` means the root of a hierarchy, and I would expect the single top level client to have the name Root, not a subclient of it.  Is it possible to rename these to illustrate that?

File: doc/Data Plane Code Generation/Subclients.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    public partial class RootClient
+    {
+        public RootClient(string cachedParameter, AzureKeyCredential credential, Uri endpoint = null, SubClientsClientOptions options = null){}
+    }
+}
+
+//Generated\ParameterClient.cs
+namespace Azure.Service.SubClients
+{
+    public partial class ParameterClient
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/Subclients.md:36]: I think it would be more helpful to our service team partners if we gave an example of clients that are named more like service functionality.  "ParameterClient" is the name of a client that tests parameters are generated, not a client that calls an Azure service.

File: doc/Data Plane Code Generation/Subclients.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+             endpoint ??= new Uri("http://localhost:3000");
+-            options ??= new SubClientsClientOptions();
++            options ??= new RootClientOptions();
+
+             ClientDiagnostics = new ClientDiagnostics(options);
+             _keyCredential = credential;
+             
+         }
+
++        private Parameter _cachedParameter0;
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/Subclients.md:221]: It's kind of confusing that this appears in the diff.  My preference would be to show just the API differences and not the implementation details.

File: doc/Data Plane Code Generation/Subclients.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+-       public ParameterClient(AzureKeyCredential credential, Uri endpoint = null, SubClientsClientOptions options = null){}
++       internal Parameter(ClientDiagnostics clientDiagnostics, HttpPipeline pipeline, AzureKeyCredential keyCredential, Uri endpoint) {}
+    }
+}
+
+```
+
+</details>
+
+### Client hierarchy
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/Subclients.md:151]: This makes Root the single top level client.  Why would I use one approach vs. another?

Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:151]: Use 'single top-level client' approach, it will use the service client (SubClientsClient) as the single top-level client, other clients are its sub-clients.  
Use 'Client hierarchy' approach, you can select one client as the top-level client, others are sub-clients according to the client hierarchy defined in customization.cs

File: doc/Data Plane Code Generation/Subclients.md [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        protected SubClientsClient()
+        {
+        }
+
+        /// <summary> Initializes a new instance of SubClientsClient. </summary>
+        /// <param name="credential"> A credential used to authenticate to an Azure Service. </param>
+        /// <param name="endpoint"> server parameter. </param>
+        /// <param name="options"> The options for configuring the client. </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="credential"/> is null. </exception>
+        public SubClientsClient(AzureKeyCredential credential, Uri endpoint = null, SubClientsClientOptions options = null)
```
[the remark this answers is not in the document]
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:88]: I am using the latest autorest.csharp, and the enpoint is optional.

File: doc/Data Plane Code Generation/Subclients.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        {
+            Argument.AssertNotNull(credential, nameof(credential));
+            endpoint ??= new Uri("http://localhost:3000");
```
[the remark this answers is not in the document]
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:91]: the endpoint url is defined in the swagger 'host', I will update the json to use an more reasonable faked url.

File: doc/Data Plane Code Generation/Subclients.md (hunk 4 above)
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:55]: No. It is default to use the ServiceClient as the to-level client, the service client name is the title of the client

Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:55]: And I don't find a way to highlight a line in an ymal code snippet. @annelo-msft Do you know any way to do so? Thanks

File: doc/Data Plane Code Generation/Subclients.md (hunk 5 above)
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:125]: I updated to the Client1Client, and Client2Client to sample the clients in the PR

File: doc/Data Plane Code Generation/Subclients.md (hunk 6 above)
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:36]: I updated to Client2Client

File: doc/Data Plane Code Generation/Subclients.md (hunk 7 above)
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:221]: removed

File: doc/Data Plane Code Generation/Subclients.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+With multiple service clients it may be better to group them all under one top-level client.
+
+<details>
+
+**Generated code before:**
+
+``` C#
+//Generated\Client1Client.cs
+namespace Azure.Service.SubClients
```
Comment (lirenhe) [inline doc/Data Plane Code Generation/Subclients.md:25]: Can we use a real service like Purview for this case?  I feel a bit difficult to understand the concept here like
the name space is call Azure.Service.SubClients, but the Client1Client should be a service client. And that name is also a bit strange

Comment (lirenhe) [inline doc/Data Plane Code Generation/Subclients.md:25]: Can we use a real service like Purview for this case?  I feel a bit difficult to understand the concept here like
the name space is call Azure.Service.SubClients, but the Client1Client should be a service client. And that name is also a bit strange

File: doc/Data Plane Code Generation/Subclients.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
+# Customizing the generated code
+
+## Subclients
+
+There are two categories of clients: service clients and their subclients. Service clients can be instantiated and have the Client suffix. Subclients can only be created by calling factory methods on other clients (commonly on service clients) and do not have the client suffix. The service client is the entry point to the API for an Azure service. Please refer to [guideline](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-subclients) for more detail information.
+
+There are two tools to organize subclients and service client.
```
Comment (lirenhe) [inline doc/Data Plane Code Generation/Subclients.md:7]: Do we provide tools to organize subclients and service client?

Comment (lirenhe) [inline doc/Data Plane Code Generation/Subclients.md:7]: Do we provide tools to organize subclients and service client?

File: doc/Data Plane Code Generation/Subclients.md [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+}
+```
+
+**Generated code after:**
+
+```diff
+namespace Azure.Service.SubClients
+{
+-   public partial class SubClientsClientOptions : ClientOptions{}
++   public partial class Client1ClientOptions : ClientOptions{}
```
Comment (lirenhe) [inline doc/Data Plane Code Generation/Subclients.md:199]: Is there any configuration change that would cause this diff in the generated code?

File: doc/Data Plane Code Generation/Subclients.md (hunk 11 above)
Comment (annelo-msft) [reply doc/Data Plane Code Generation/Subclients.md:25]: I would love this!  I would love if we could start thinking in terms of the impact generator changes have to service libraries more than just the autorest.csharp test output.  The test client output helps us explore edge cases and get a preview of what updates to the generator will look like in the azure-sdk-for-net repo, but they are hard to understand outside of the autorest context, and we shouldn't expect service team partners to try to learn these.

File: doc/Data Plane Code Generation/Subclients.md (hunk 4 above)
Comment (annelo-msft) [reply doc/Data Plane Code Generation/Subclients.md:55]: See https://github.com/Azure/azure-sdk-for-net/pull/27559#discussion_r838820230

Comment (annelo-msft) [reply doc/Data Plane Code Generation/Subclients.md:55]: I don't know for yaml specifically - is this something you could investigate?

**Review by annelo-msft — COMMENTED** [review]
I will review this again once it has been changed to a client for a service library, as @lirenhe  suggested.

File: doc/Data Plane Code Generation/Subclients.md (hunk 11 above)
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:25]: Hi @annelo-msft  @lirenhe  we are familiar with some real service like Purview, in that case, it may be more easy to understand if we take Puerview as example. But actually the reader of the document will have no ideal about Purview, and the complex of the Purview service may make the document more difficult to understand. The document is to show subclient feature, I think it is better to use a sample RP with simple client structure, then it is more easy to show the subclient principle.  By the way, customer need to understand what is subclient then he can understand the document, so in the beginning we add subclient guideline reference link.

Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:25]: same as above

File: doc/Data Plane Code Generation/Subclients.md (hunk 12 above)
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md:7]: @lirenhe  what tools do you meantion?

File: doc/Data Plane Code Generation/Subclients.md [hunk 14]
```diff
@@ -0,0 +1,233 @@
+# Customizing the generated code
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/Subclients.md:1]: It may be worth pulling in information from this document as well: https://github.com/Azure/autorest.csharp/blob/feature/v3/docs/features/subclients.md

We may need to bring it up to date, but it provides a nice description.

File: doc/Data Plane Code Generation/Subclients.md [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+## Subclients
```
Comment (annelo-msft) [inline doc/Data Plane Code Generation/Subclients.md:3]: It would be good to give guidance on how a partner could create multiple service clients as well.  Is the guidance for this just to put them in separate operation groups?

File: doc/Data Plane Code Generation/Subclients.md (hunk 12 above)
Comment (lirenhe) [reply doc/Data Plane Code Generation/Subclients.md:7]: I saw this sentense  'There are two tools to organize subclients and service client'

File: doc/Data Plane Code Generation/Subclients.md (hunk 11 above)
Comment (lirenhe) [reply doc/Data Plane Code Generation/Subclients.md:25]: I don't think user need to understand Purview to understand the subclient. For me, the name like 'Client1Client' make me feel confused.  The PR https://github.com/Azure/azure-sdk-for-net/pull/27558  also use a MetricAdvisor as an example and I didn't see any difficultiy to understand it.

File: doc/DataPlaneCodeGeneration/Subclients.md [hunk 16]
```diff
@@ -0,0 +1,295 @@
+# Customizing the generated code
+
+## Subclients
+
+There are two categories of clients: service clients and their subclients. Service clients can be instantiated and have the Client suffix. Subclients can only be created by calling factory methods on other clients (commonly on service clients) and do not have the client suffix. The service client is the entry point to the API for an Azure service. Please refer to [guideline](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-subclients) for more detail information.
+
+There are two tools to organize subclients and service client.
```
Comment (lirenhe) [inline doc/DataPlaneCodeGeneration/Subclients.md:7]:
```suggestion
There are two ways to organize subclients and service client.
```

File: doc/DataPlaneCodeGeneration/Subclients.md [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+### AutoRest Configuration
+> see https://aka.ms/autorest
+
+``` yaml
+namespace: Azure.Service.SubClients
+input-file: $(this-folder)/SubClients.json
+data-plane: true
+security: AzureKey
+security-header-name: Fake-Subscription-Key
+single-top-level-client: true
```
Comment (lirenhe) [inline doc/DataPlaneCodeGeneration/Subclients.md:66]: Seems this is the key setting that makes the top level client. Is there a way to highlight that?

File: doc/DataPlaneCodeGeneration/Subclients.md [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+{
+    public partial class ResourceSetRulesClient
+    {
+        public ResourceSetRulesClient(string endpoint, TokenCredential credential, PurviewAccountClientOptions options = null){}
+    }
+}
+```
+
+**Add client hierarchy customization (Customizations.cs):**
+
```
Comment (lirenhe) [inline doc/DataPlaneCodeGeneration/Subclients.md:226]: Curious about the steps here. Do i need to first generate the original code, add Customizations.cs into the source code folder and generate again?

**Comment by Stordlord** [comment]
APPROVE

File: doc/DataPlaneCodeGeneration/Subclients.md (hunk 18 above)
Comment (chunyu3) [reply doc/DataPlaneCodeGeneration/Subclients.md:226]: > Curious about the steps here. Do i need to first generate the original code, add Customizations.cs into the source code folder and generate again?

Without first generating the original code, you will not know which clients are in the PR and write the Customization.cs to setup the hierarchy. If you can figure out the clients and set the client hierarchy, you will not generate the original code.

File: doc/DataPlaneCodeGeneration/Subclients.md (hunk 17 above)
Comment (chunyu3) [reply doc/DataPlaneCodeGeneration/Subclients.md:66]: I have not find a way to highlight in ymal code snippet in a md file.

File: doc/DataPlaneCodeGeneration/Subclients.md (hunk 16 above)
Comment (chunyu3) [reply doc/DataPlaneCodeGeneration/Subclients.md:7]: resolved

File: doc/Data Plane Code Generation/Subclients.md (hunk 11 above)
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md]: update to use a real service Purview Account for example.

File: doc/Data Plane Code Generation/Subclients.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (chunyu3) [reply doc/Data Plane Code Generation/Subclients.md]: It is because client1 is promoted as top-level service client. the clientOptions name changed accordingly

File: doc/DataPlaneCodeGeneration/Subclients.md [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+  },
+  "/nogroup/op": {
+    "get": {
+      "operationId": "Op2",
+    }
+  }
+}
+```
+
+```yaml
```
Comment (lirenhe) [inline doc/DataPlaneCodeGeneration/Subclients.md:22]: Is this modeler4 code?  It could be difficult for people who do not know M4 to understand this part.

Comment (lirenhe) [reply doc/DataPlaneCodeGeneration/Subclients.md:22]: I suggest we could remove that. WIthout this, i could still understand the content.

File: doc/DataPlaneCodeGeneration/Subclients.md (hunk 17 above)
Comment (lirenhe) [reply doc/DataPlaneCodeGeneration/Subclients.md:127]: Then could you highlight that in the description under 
**Add customize configuration (set single-top-level-client):**

**Review by lirenhe — APPROVED** [review]
LGTM

File: doc/DataPlaneCodeGeneration/Subclients.md [hunk 20]
```diff
@@ -0,0 +1,332 @@
+# Customizing the generated code
+
+## Subclients
+
+autorest requires each operation in swagger file to define a unique string parameter `operationId` (this is different from [official OpenAPI specification](https://swagger.io/docs/specification/paths-and-operations/#operationId)). If parameter value contains underscore, then the part after underscore will be treated as operation name, and the part before underscore will be treated as the name of the operation group which this operation belongs to. Otherwise, operation will be attributed to the group without a name::
```
Comment (pshao25) [inline doc/DataPlaneCodeGeneration/Subclients.md:5]: Autorest

**Review by pshao25 — APPROVED** [review]

**Comment by chunyu3** [comment]
@annelo-msft would you please help to review the document? thanks

**Review by annelo-msft — COMMENTED** [review]
This is looking great!

One more comment: I think it's easier for partners to understand documents that "start with why" than those that jump straight to the technical implementation.  It would be great if we could add a sentence or two at the beginning explaining why we use subclients, i.e. why would someone want to follow these instructions.  You should be able to get this information from the .NET Guidelines, let me know if you need help interpreting it into a short introduction here.

@lmazuel, do you have any concerns with focusing on operation id with regard to subclients in this doc?

**Comment by chunyu3** [comment]
> This is looking great!
> 
> One more comment: I think it's easier for partners to understand documents that "start with why" than those that jump straight to the technical implementation. It would be great if we could add a sentence or two at the beginning explaining why we use subclients, i.e. why would someone want to follow these instructions. You should be able to get this information from the .NET Guidelines, let me know if you need help interpreting it into a short introduction here.
> 
> @lmazuel, do you have any concerns with focusing on operation id with regard to subclients in this doc?

Hi @annelo-msft , I added some "start with why" describe as following:
As above, one Azure service maybe be exposed to .NET developers as more than one clients that is not easy for .NET developers to discover and find out the right starting point to call Azure services with the Azure SDK. To resolve this, we can export only one service client in its main namespace as the main starting point for developers calling Azure services with the Azure SDK via [subclient feature](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-subclients).

File: doc/DataPlaneCodeGeneration/Subclients.md [hunk 21]
```diff
@@ -0,0 +1,335 @@
...
+
+```cs
+public partial class MyServiceClient
+{
+    public virtual async Task<Response> Op2Async(RequestContext options = null);
+    public virtual Response Op2(RequestContext options = null);
+}
+```
+
+As above, one Azure service maybe be exposed to .NET developers as more than one clients that is not easy for .NET developers to discover and find out the right starting point to call Azure services with the Azure SDK. To resolve this, we can export only one service client in its main namespace as the main starting point for developers via [subclient feature](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-subclients).
```
Comment (annelo-msft) [inline doc/DataPlaneCodeGeneration/Subclients.md:40]: Please put this before line 5.

**Review by annelo-msft — APPROVED** [review]

File: doc/DataPlaneCodeGeneration/Subclients.md (hunk 21 above)
Comment (chunyu3) [reply doc/DataPlaneCodeGeneration/Subclients.md:40]: Those statements is based on the information of line 1-39. So I think it is not good to put before line 5.
